### ISSUE TITLE
wasm2c load/store offset calculation

### ISSUE THREAD

**Issue by kripken** [WebAssembly/wabt#1400]
Fuzz testcase:
```wat
 (module
 (type $none_=>_none (func))
 (memory $0 1 1)
 (global $global$0 (mut i32) (i32.const 10))
 (export "hangLimitInitializer" (func $1))
 (export "func_13_invoker" (func $0))
 (func $0
  (loop $label$1
   (if
    (global.get $global$0)
    (block
     (global.set $global$0
      (i32.const 0)
     )
     (drop
      (i32.load8_s offset=22  ;; note the offset is bigger than 16
       (i32.const -16)
      )
     )
     (br $label$1)
    )
   )
  )
 )
 (func $1
  (nop)
 )
)
```
The key part is
```wat
      (i32.load8_s offset=22  ;; note the offset is bigger than 16
       (i32.const -16)
      )
```
Wasm2c turns that into
```c
      i0 = 4294967280u;
      i0 = i32_load8_s((&M0), (u64)(i0 + 22));
```
The wasm here would trap, but the wasm2c output doesn't. I think what happens is `i0 + 22` wraps around, and only then do we turn it into a `u64` and do a load from the seemingly-valid address `6`.

**Comment by binji** [comment]
Ah, good point. It should cast the base to (u64) before adding the offset. I'm kinda surprised there isn't a spec test that would catch this case...

**Comment by binji** [comment]
There is a spec test that should have caught this in `address.wast`:

```wasm
(module
  (memory 1)
  ...
  (func (export "8u_bad") (param $i i32)
    (drop (i32.load8_u offset=4294967295 (local.get $i)))
  )
)

(assert_trap (invoke "8u_bad" (i32.const 1)) "out of bounds memory access")
```

Which generates:

```c
void _u_bad(u32 p0) {
  FUNC_PROLOGUE;
  u32 i0;
  i0 = p0;
  i0 = i32_load8_u((&M0), (u64)(i0 + 4294967295));
  FUNC_EPILOGUE;
}
```

Which should have the same issue, but it seems the generated code always traps (see on godbolt for example: https://godbolt.org/z/U-HCjc). Not sure why that would be.

**Comment by binji** [comment]
Ah I see. It's because the type of `4294967295` is long, so `uint32_t + long => long`, which is 8 bytes.

But when the offset is smaller than 2<sup>31</sup> then it is an `int`, so the sum is `uint32_t + int => uint32_t`, which is 4 bytes and will wrap.

**Comment by binji** [comment]
See new tests here: https://github.com/WebAssembly/spec/pull/1188

**Comment by kripken** [comment]
Looks like this was fixed by #1401 !
